File: leetcode/solutions/0387.FirstUniqueCharacterInAString/firstUniqChar.go
package solution

func firstUniqChar(s string) int {
	a := 'a'
	firstCharIndices := make([]int, 26)
	distincts := make([]bool, len(s))

	for i, r := range s {
		j := r - a
		firstIndex := firstCharIndices[j]

		if firstIndex != 0 {
			distincts[firstIndex-1] = false
		} else {
			firstCharIndices[j] = i + 1
			distincts[i] = true
		}
	}

	for i, isDistinct := range distincts {
		if isDistinct {
			return i
		}
	}

	return -1
}

File: leetcode/solutions/0387.FirstUniqueCharacterInAString/firstUniqChar.ts
export default function firstUniqChar(s: string): number {
  const firstCharIndices = Array.from({ length: 26 }, () => 0);
  const distinct = Array.from({ length: s.length }, () => false);
  const a = "a".charCodeAt(0);

  for (let i = 0; i < s.length; i++) {
    const j = s.charCodeAt(i) - a;
    const firstIndex = firstCharIndices[j];

    if (firstIndex === 0) {
      firstCharIndices[j] = i + 1;
      distinct[i] = true;
    } else {
      distinct[firstIndex - 1] = false;
    }
  }

  for (const [i, isDistinct] of distinct.entries()) {
    if (isDistinct) {
      return i;
    }
  }
  return -1;
}
